File: app/src/main/java/com/checkin/app/checkin/Data/NetworkBoundResource.java
package com.checkin.app.checkin.Data;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MediatorLiveData;
import android.os.AsyncTask;
import android.support.annotation.MainThread;
import android.support.annotation.NonNull;
import android.support.annotation.WorkerThread;
import android.util.Log;

// ResultType: Type for the Resource data
// RequestType: Type for the API response
public abstract class NetworkBoundResource<ResultType, RequestType> {
    private final static String TAG = NetworkBoundResource.class.getSimpleName();
    private final MediatorLiveData<Resource<ResultType>> mResult = new MediatorLiveData<>();

    @MainThread
    public NetworkBoundResource() {
        mResult.setValue(Resource.loading(null));
        if (shouldUseLocalDb()) {
            final LiveData<ResultType> dbSource = loadFromDb();
            mResult.addSource(dbSource, data -> {

                mResult.removeSource(dbSource);
                if (shouldFetch(data)) {
                    fetchFromNetwork(dbSource);
                } else {
                    mResult.addSource(dbSource,
                            (ResultType newData) -> mResult.setValue(Resource.success(newData)));
                }
            });
        } else {
            fetchFromNetwork(null);
        }
    }

    private void fetchFromNetwork(LiveData<ResultType> dbSource) {
        LiveData<ApiResponse<RequestType>> apiResponse = createCall();
        final boolean useDb = shouldUseLocalDb() && dbSource != null;
        if (useDb) {
            // we re-attach dbSource as a new source,
            // it will dispatch its latest value quickly
            mResult.addSource(dbSource, newData -> {
                mResult.setValue(Resource.loading(newData));
            });
        }
        mResult.addSource(apiResponse, response -> {
            mResult.removeSource(apiResponse);
            if (useDb) {
                mResult.removeSource(dbSource);
            }
            if (response != null) {
                Resource<RequestType> resource = Resource.createResource(response);
                if (useDb)
                    saveResultAndReInit(resource);
                else
                    postResultDirectly(resource);
                if (!response.isSuccessful()) {
                    onFetchFailed(response);
                    if (useDb) {
                        mResult.addSource(dbSource, newData -> {
                            mResult.setValue(Resource.error(resource.message, newData));
                        });
                    }
                }
            }
        });
    }

    @MainThread
    private void saveResultAndReInit(Resource<RequestType> resource) {
        new AsyncTask<Void, Void, Void>() {

            @Override
            protected Void doInBackground(Void... params) {
                saveCallResult(resource.data);
                return null;
            }

            @Override
            protected void onPostExecute(Void param) {
                // we specially request a new live data,
                // otherwise we will get immediately last cached value,
                // which may not be updated with latest results received from network.
                mResult.addSource(loadFromDb(),
                        newData -> mResult.setValue(Resource.success(newData)));
            }
        }.execute();
    }

    // Called in case no Database interaction needed.
    @MainThread
    protected void postResultDirectly(Resource<RequestType> resource) {
        RequestType data = resource.data;
        Log.e(TAG, resource.status.name());
        try {
            mResult.setValue(Resource.cloneResource(resource, (ResultType) data));
        } catch (ClassCastException e) {
            Log.e(TAG, "Invalid Resource Data type.");
        }
    }

    // Called with the data in the database to decide whether it should be
    // fetched from the network.
    @MainThread
    protected boolean shouldFetch(ResultType data) {
        return true;
    }

    protected abstract boolean shouldUseLocalDb();

    // Called to create the API call.
    @NonNull
    @MainThread
    protected abstract LiveData<ApiResponse<RequestType>> createCall();

    // Called to get the cached data from the database
    @MainThread
    protected LiveData<ResultType> loadFromDb() {
        return null;
    }

    // Called to save the result of the API response into the database
    @WorkerThread
    protected abstract void saveCallResult(RequestType data);

    // Called when the fetch fails. The child class may want to reset components
    // like rate limiter.
    @MainThread
    protected void onFetchFailed(ApiResponse<RequestType> response) {
    }

    // returns a LiveData that represents the resource, implemented
    // in the base class.
    public LiveData<Resource<ResultType>> getAsLiveData() {
        return mResult;
    }
}

File: app/src/main/java/com/checkin/app/checkin/data/resource/NetworkBoundResource.kt
package com.checkin.app.checkin.data.resource

import android.util.Log
import androidx.annotation.MainThread
import androidx.annotation.WorkerThread
import androidx.lifecycle.LiveData
import androidx.lifecycle.MediatorLiveData
import com.checkin.app.checkin.data.network.ApiResponse
import com.checkin.app.checkin.data.resource.Resource.Companion.cloneResource
import com.checkin.app.checkin.data.resource.Resource.Companion.createResource
import com.checkin.app.checkin.data.resource.Resource.Companion.errorButLoadedCached
import com.checkin.app.checkin.data.resource.Resource.Companion.errorNotFoundCached
import com.checkin.app.checkin.data.resource.Resource.Companion.loading
import com.checkin.app.checkin.data.resource.Resource.Companion.success
import com.checkin.app.checkin.data.resource.Resource.Companion.successCached
import com.checkin.app.checkin.utility.Utils
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking

// ResultType: Type for the Resource data
// RequestType: Type for the API response
abstract class NetworkBoundResource<ResultType, RequestType> : CoroutineScope by CoroutineScope(Dispatchers.IO) {
    private val mResult = MediatorLiveData<Resource<ResultType>>()

    // returns a LiveData that represents the resource, implemented
    // in the base class.
    val asLiveData: LiveData<Resource<ResultType>> = mResult

    init {
        fetchData()
    }

    private fun fetchFromNetwork(dbSource: LiveData<ResultType>?) {
        val apiResponse = createCall()!!
        if (dbSource != null) {
            // we re-attach dbSource as a new source, it will dispatch its latest value quickly
            mResult.addSource(dbSource) {
                mResult.removeSource(dbSource)
                mResult.value = loading(it)
            }
        }
        mResult.addSource(apiResponse) { response ->
            mResult.removeSource(apiResponse)
            if (response != null) {
                val resource = createResource(response)
                if (shouldUseLocalDb() && resource.status == Resource.Status.SUCCESS) saveResultAndReInit(resource)
                else postResultDirectly(resource, response.requestUrl)
                if (!response.isSuccessful) {
                    onFetchFailed(response)
                    if (dbSource != null) {
                        mResult.addSource(dbSource) {
                            if (it == null || (it is List<*> && it.isEmpty())) mResult.value = errorNotFoundCached<ResultType>(resource.message)
                            else mResult.value = errorButLoadedCached(resource.message, it)
                        }
                    }
                }
            }
        }
    }

    protected open fun fetchData() {
        mResult.value = loading(null)
        val dbSource = if (shouldUseLocalDb()) loadFromDb() else null
        if (dbSource != null) {
            mResult.addSource(dbSource) { data ->
                mResult.removeSource(dbSource)
                if (shouldFetch(data)) {
                    fetchFromNetwork(dbSource)
                } else {
                    mResult.addSource(dbSource) {
                        if (it == null || (it is List<*> && it.isEmpty())) mResult.value = errorNotFoundCached<ResultType>(null)
                        else mResult.value = successCached(it)
                    }
                }
            }
        } else fetchFromNetwork(null)
    }

    @MainThread
    private fun saveResultAndReInit(resource: Resource<RequestType>) = runBlocking {
        launch { saveCallResult(resource.data) }.join()
        // we specially request a new live data,
        // otherwise we will get immediately last cached value,
        // which may not be updated with latest results received from network.
        val dbResponse = loadFromDb()
        if (dbResponse != null) mResult.addSource(dbResponse) {
            mResult.removeSource(dbResponse)
            if (it != null) mResult.value = transformResult(resource, success(it))
            else mResult.value = cloneResource(resource, null)
        } else mResult.value = transformResult(resource)
    }

    // Called in case no Database interaction needed.
    @MainThread
    protected fun postResultDirectly(resource: Resource<RequestType>, url: String?) {
        Log.d(TAG, "${resource.status.name} - $url")
        mResult.value = transformResult(resource)
    }

    /**
     * Transforms the network resource received at the end of lifecycle
     * By default, if [dbResult] is present returns that otherwise returns the network resource, [networkResult]
     */
    protected open fun transformResult(networkResult: Resource<RequestType>, dbResult: Resource<ResultType>? = null): Resource<ResultType> = kotlin.runCatching {
        dbResult ?: cloneResource(networkResult, networkResult.data as? ResultType)!!
    }.getOrElse {
        Utils.logErrors(TAG, it, "Invalid Resource Data type.")
        throw it
    }

    // Called with the data in the database to decide whether it should be
    // fetched from the network.
    @MainThread
    protected open fun shouldFetch(data: ResultType?): Boolean = true

    protected open fun shouldUseLocalDb(): Boolean = false

    // Called to create the API call.
    @MainThread
    protected abstract fun createCall(): LiveData<ApiResponse<RequestType>>?

    // Called to get the cached data from the database
    @MainThread
    protected open fun loadFromDb(): LiveData<ResultType>? = null

    // Called to save the result of the API response into the database
    @WorkerThread
    protected open fun saveCallResult(data: RequestType?) {
    }

    // Called when the fetch fails. The child class may want to reset components
    // like rate limiter.
    @MainThread
    protected fun onFetchFailed(response: ApiResponse<RequestType>?) {
    }

    companion object {
        private val TAG = NetworkBoundResource::class.java.simpleName
    }
}